src/coreclr/jit/fgehopt.cpp

```diff
@@ -2097,3 +2097,649 @@ PhaseStatus Compiler::fgTailMergeThrows()
+                {
+                    addBlockToClone(block, "callfinally");
+                }
+                else if (block->KindIs(BBJ_CALLFINALLYRET) && block->Prev()->TargetIs(ebd->ebdHndBeg))
+                {
+                    addBlockToClone(block, "callfinallyret");
+                }
+            }
+        }
+
+        // Walk the filter region
+        //
+        if (ebd->HasFilter())
+        {
+            BasicBlock* const firstFltBlock = ebd->ebdFilter;
+            BasicBlock* const lastFltBlock  = ebd->BBFilterLast();
+
+            if (BitVecOps::IsMember(traits, visited, firstFltBlock->bbNum))
+            {
+                JITDUMP("already walked filter region for EH#%02u\n", regionIndex);
+                assert(BitVecOps::IsMember(traits, visited, lastFltBlock->bbNum));
+            }
+            else
+            {
+                JITDUMP("walking filter region for EH#%02u\n", regionIndex);
+                for (BasicBlock* const block : Blocks(firstFltBlock, lastFltBlock))
+                {
+                    // A filter cannot enclose another EH region
+                    //
+                    assert(!bbIsTryBeg(block));
+                    addBlockToClone(block, "filter region");
+                }
+            }
+        }
+
+        // Walk the handler region
+        //
+        BasicBlock* const firstHndBlock = ebd->ebdHndBeg;
+        BasicBlock* const lastHndBlock  = ebd->ebdHndLast;
+
+        if (BitVecOps::IsMember(traits, visited, firstHndBlock->bbNum))
+        {
+            JITDUMP("already walked handler region for EH#%02u\n", regionIndex);
+            assert(BitVecOps::IsMember(traits, visited, lastHndBlock->bbNum));
+        }
+        else
+        {
+            JITDUMP("walking handler region for EH#%02u\n", regionIndex);
+            for (BasicBlock* const block : Blocks(firstHndBlock, lastHndBlock))
+            {
+                bool added = addBlockToClone(block, "handler region");
+                if (bbIsTryBeg(block))
+                {
+                    assert(added);
+                    JITDUMP("==> found try entry for EH#%02u nested in handler at " FMT_BB "\n", block->bbNum,
+                            block->getTryIndex());
+                    regionsToProcess.Push(block->getTryIndex());
+                }
+            }
+        }
+
+        // If there is an enclosing mutual-protect region, process it as well
+        //
+        unsigned const enclosingTryIndex = ebd->ebdEnclosingTryIndex;
+        if (enclosingTryIndex != EHblkDsc::NO_ENCLOSING_INDEX)
+        {
+            EHblkDsc* const enclosingTryEbd = ehGetDsc(enclosingTryIndex);
+
+            if (EHblkDsc::ebdIsSameTry(ebd, enclosingTryEbd))
+            {
+                JITDUMP("==> found mutual-protect try EH#%02u for EH#%02u\n", enclosingTryIndex, regionIndex);
+                regionsToProcess.Push(enclosingTryIndex);
+            }
+        }
```

Comment:
Can there be multiple of these?

Reply:
Yes, there is one per catch in a try/catch/.../catch. Each enclosed try adds its enclosing try, so we get them all.